File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.add(c);
            } else {
                if  ( !stack.isEmpty() && 
                    ((c == ')' && stack.peek()=='(') || 
                    (c == '}' && stack.peek()=='{') || 
                    (c == ']' && stack.peek()=='['))) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapping = { ")": "(", "}":"{", "]":"["}
        stack = []
        for c in s:
            if not c in mapping:
                stack.append(c)
            else:
                if len(stack) > 0 and stack[-1] == mapping[c]:
                    stack.pop()
                else:
                    return False
        
        return len(stack)==0
            